### ISSUE TITLE
Language model

### ISSUE THREAD

**Issue by rajasumanth** [SeanNaren/deepspeech.pytorch#149]
I am using Libri speech language model which is not giving me good results, could you please suggest how to acquire a good language model or create a good language model

Thanks in advance

**Comment by jinserk** [comment]
I've tested two public language models: cantab-research model and kenLM binary model from Mozilla's [DeepSpeech](https://github.com/Mozilla/DeepSpeech.git). In order to use both of them, you need to compile KenLM frorm [here](https://github.com/kpu/kenlm.git)

**Comment by SeanNaren** [comment]
I'm not seeing expected results using language models with pytorch-ctc. Investigating further!

**Comment by SeanNaren** [comment]
You have to tweak a few parameters. I'd suggest starting by adding the parameter `--lm_alpha 2.0` and increasing the `--lm_beta1` slightly and seeing how it performs

**Comment by acc8518** [comment]
Hello! Thank you for providing such good codes!

I'm also trying to train a 3-gram model with [Libri speech data](http://www.openslr.org/resources/11/librispeech-lm-norm.txt.gz) using the Kenlm tool . I use the language model to do the experiment with the [LibriSpeech pretrained model](https://github.com/SeanNaren/deepspeech.pytorch/releases) and the libri_val_manifest.csv, provided by this code. But the transcript results always full of blank or " ' ". The result is good without using the beam search with the language model.
I also try to tune the parameters of lm_alpha and lm_beta as mentioned, but it seems not work to me.
Would me mind to suggest me how to get a better result?

**Comment by saharghannay** [comment]
Hi, 
Did you please tried the  beam search decoding with langage model, using the updated  version of ctcdecode?
Because, I tried and I did not succeed, I had segmentation fault problem. 

Thank you,

**Comment by acc8518** [comment]
Hi,

FYI. I got the same error when I used the Libri LM downloaded from
the Openslr, which is an .arpa file. I can successfully load it after I
transform it to the binary format.

However, this method cannot work when I used the Cantab-TEDLIUM Release
1.1, and then I downloaded the corpus and trained the LM on my computer, it
worked.

Hope it helps you.

2017-12-04 22:39 GMT+08:00 saharghannay <notifications@github.com>:

**Comment by saharghannay** [comment]
Hi,
@acc8518  did you use a trie file ?
 The lm_path option is it for the lm file or the trie file ?
Because I am using this command line and I am not sure if is it correct.
 python test.py  —model_path models/deepspeech_final.pth.tar —cud —test_manifest ted_test_manifest.csv  —decoder beam  —output_path models/output --lm_path cantab-TEDLIUM-pruned.lm3.bin

Thank you,

**Comment by acc8518** [comment]
hi,

I did not use a trie file. The lm_path option is for the lm file.

**Comment by saharghannay** [comment]
Ok, 
I tried to build a new LM with Kenlm ( bin/lmplz -o 3 <vocab.txt > 3glm.arpa), but I still have the same problem.
Did you change the labels.json file ?

**Comment by acc8518** [comment]
Hi,

No, I did not change the label file.

**Comment by saharghannay** [comment]
Hi,
Ok, did you use specific options to binaries the arpa file ?
I am using the standard one: build_binary 3glm.arpa 3glm.bin

Thank you

**Comment by acc8518** [comment]
Hi,

No, I did not use specific options.

2017-12-05 18:50 GMT+08:00 saharghannay <notifications@github.com>:

**Comment by ryanleary** [comment]
Closing on lack of activity.
